internal/bootstrap/controller.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package bootstrap
+
+import (
+	"context"
+
+	"github.com/juju/errors"
+)
+
+const controllerCharmURL = "juju-controller"
+
+// PopulateControllerCharm is the function that is used to populate the
```

Comment:
Comment might better describe acquiring and deploying the charm, and using it to add the application and unit.
Pretty much what you have inside the function, could constitute its comment at the top instead.